### PR TITLE
Data validations and unit tests

### PR DESCRIPTION
**Pull request by Reggeditt**
## In this pull request, I;

- Add the following validations:
    - For the User model:
Name must not be blank.
PostsCounter must be an integer greater than or equal to zero.
    - For the Post model:
Title must not be blank.
Title must not exceed 250 characters.
CommentsCounter must be an integer greater than or equal to zero.
LikesCounter must be an integer greater than or equal to zero.

- Add unit specs for all models' methods and validations.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,2 +1,11 @@
 class User < ApplicationRecord
+  has_many :posts
+  has_many :comments
+  has_many :likes
+  validates :name, presence: true, length: { minimum: 3, maximum: 20 }
+  validates :posts_counter, numericality: { greater_than_or_equal_to: 0 }
+
+  def recent_posts(limit = 3)
+    posts.order(created_at: :desc).limit(limit)
+  end
```
Comment (nahnahdev) [inline app/models/user.rb]: - [ ] Great job so far @Reggeditt  There are just a few changes I will suggest you implement and you'll be on your way to the next milestone :+1: 
- [ ] Per the project requirements, you were asked to `Add unit specs for all of your models' methods and validations` You have done an amazing job adding most of the tests. I noticed that this `recent_posts` method hasn't been tested yet. So kindly include that in your tests :wink:

File: app/models/post.rb [hunk 2]
```diff
@@ -1,3 +1,24 @@
...
+    update(comments_counter: comments.count)
+  end
+
+  def update_likes_counter
+    update(likes_counter: likes.count)
+  end
+
+  def recent_comments(limit = 5)
+    comments.order(created_at: :desc).limit(limit)
+  end
```
Comment (nahnahdev) [inline app/models/post.rb]: - [ ] Kindly add `unit specs` for this `recent_comments` method as well :+1:

File: app/models/post.rb [hunk 3]
```diff
@@ -1,3 +1,24 @@
...
+  validates :comments_counter, numericality: { greater_than_or_equal_to: 0 }
+  validates :likes_counter, numericality: { greater_than_or_equal_to: 0 }
+
+  def update_comments_counter
+    update(comments_counter: comments.count)
+  end
+
+  def update_likes_counter
+    update(likes_counter: likes.count)
+  end
```
Comment (nahnahdev) [inline app/models/post.rb]: - [ ] Kindly add `unit specs` for this `update_comments_counter` and `update_likes_counter` method as well :+1:

File: spec/models/post_spec.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  it 'title should not exceed 250 characters' do
+    post = Post.new(title: 'a' * 251)
+    expect(post).to_not be_valid
+  end
+
+  it 'text should be present' do
+    post = Post.new(text: nil)
+    expect(post).to_not be_valid
+  end
+end
```
Comment (nahnahdev) [inline spec/models/post_spec.rb]: - [ ] I also noticed that your tests don't contain tests for these two validations below in your Post Model in the  `post.rb` file so kindly include those as well :+1: 

```ruby
 validates :comments_counter, numericality: { greater_than_or_equal_to: 0 }
  validates :likes_counter, numericality: { greater_than_or_equal_to: 0 }

```

File: app/models/like.rb [hunk 5]
```diff
@@ -1,4 +1,14 @@
...
+
+  alias_attribute :author, :user
+
+  after_create :update_post_likes_counter
+
+  private
+
+  def update_post_likes_counter
+    post.update_likes_counter
+  end
```
Comment (nahnahdev) [inline app/models/like.rb]: - [ ] Kindly add `unit specs` for this `update_post_likes_counter` method as well :+1:

File: app/models/comment.rb [hunk 6]
```diff
@@ -1,4 +1,14 @@
...
+
+  alias_attribute :author, :user
+
+  after_create :update_post_comments_counter
+
+  private
+
+  def update_post_comments_counter
+    post.update_comments_counter
+  end
```
Comment (nahnahdev) [inline app/models/comment.rb]: - [ ] Kindly add `unit specs` for this `update_post_comments_counter` method as well :+1:

**Review by nahnahdev — CHANGES_REQUESTED** [review]
## Project Status: CHANGES REQUIRED :recycle: 
Hi @Reggeditt  , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights :fire: 
- All required user validations added :heavy_check_mark: 
- All required posts validations added :heavy_check_mark: 
- Has unit tests for `User` and `Post` model :heavy_check_mark: 
- All tests passing :heavy_check_mark: 
- Most model specs has **Validation** tests :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
- N/A

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb [hunk 7]
```diff
@@ -1,3 +1,33 @@
...
+    author.increment!(:posts_counter)
+  end
+
+  def decrement_author_posts_counter
+    author.decrement!(:posts_counter)
+  end
+
+  def update_likes_counter
+    update(likescounter: likes.count)
+  end
```
Comment (aamirkhan2478) [inline app/models/post.rb:32]: - [ ] As mentioned by the previous reviewer kindly add unit specs for the `update_likes_counter` method as well but still, you didn't add any specs for `update_likes_counter`👍

**Review by aamirkhan2478 — CHANGES_REQUESTED** [review]
Hi @Reggeditt, 



Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.

 ## Highlights
✔️ Good Effort
✔️ Your design is good.
✔️ All linter passed

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/like_spec.rb
Comment (fobadara) [inline spec/models/like_spec.rb:1]: [OPTIONAL] In future you should explain to the reviewer in a comment that you have implemented the requirement. The only difference here is that you named it differently which is allowed in this project.

**Review by fobadara — APPROVED** [review]
Hi @Reggeditt, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

APPRVED

## Highlights
✅ Descriptive README
✅ Descriptive PR summary
✅ Clean code



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**